### PR TITLE
Add multiple subscription mechanism

### LINKED ISSUE #144: Enhancing subscribe mechanism
The hunter subscribe mechanism currently support subscribing to one or multiple event.
The problem arises when for example we want to run a hunter only if two specific events were published.

One thing we could do is add a new 'queue' dictionary, that will store the 'multiple subscriptions', and on each event that is being published that resides in this queue_dict, adding the event object to the queue. 
In the case that the list of event objects corresponds to the requirement of the publish, the events will be concatenated, and the hunter hook will execute. This logic would happen each time on the publish_event() method.

### PR DESCRIPTION
**Pull request by RaitoBezarius**
## Description
As per #144 ; this PR enables a `multiple_subscribe` decorator which waits for multiple events to be fired before the hook is executed.
It also supports multiple predicates for each event.

It passes the set of `events` directly to the hook.

I guess we can refine it a lot more to fit more use cases, but I didn't want to go overkill, I believe the greatest generalization would be full logic `@complex_subscribe(SomeEvent & (EventA | ~EventB))` for example, and we could feed predicates using a dict. (where `&` means AND, `|` means OR and `~` means NOT).

That sounds cool, but in practice, is there a need for it?

## Fixed Issues

Fixes #144.

## Contribution checklist
 - [x] I have read the Contributing Guidelines.
 - [x] The commits refer to an active issue in the repository.
 - [x] I have added automated testing to cover this case.

### CODE REVIEW

File: tests/core/test_subscribe.py [hunk 1]
```diff
@@ -26,25 +26,38 @@ def __init__(self, event):
...
-    
+
+    # testing the multiple subscription mechanism
+    handler.publish_event(OnceOnlyEvent())
+
+    time.sleep(0.02)
+    # SmartHunter and OnceHunter should run only.
+    assert counter == 2
+    counter = 0
+
```
Comment (iYehuda) [inline tests/core/test_subscribe.py]: I think this code deserves its own test (i.e, "test_subscribe_many").
The current test is already overloaded with "subscribe" and "subscribe_once" testing, so you can split that too (please).

File: tests/core/test_subscribe.py [hunk 2]
```diff
@@ -26,25 +26,38 @@ def __init__(self, event):
...
     handler.publish_event(RegularEvent())
-    
+
     time.sleep(0.02)
     assert counter == 3
     counter = 0
-    
+
+    # testing the multiple subscription mechanism
+    handler.publish_event(OnceOnlyEvent())
```
Comment (iYehuda) [inline tests/core/test_subscribe.py]: Please add test case of subscriber being triggered twice.

File: src/core/events/handler.py [hunk 3]
```diff
@@ -44,6 +47,13 @@ def wrapper(hook):
 
         return wrapper
 
+    def multiple_subscribe(self, events, predicates=None):
```
Comment (iYehuda) [inline src/core/events/handler.py]: Rename to `subscribe_many` to align with `subscribe_once`

**Comment by iYehuda** [comment]
I'd like this PR to get another pair of eyes on it since it's a little complicated.
@itaysk can you review it?

**Comment by RaitoBezarius** [comment]
I addressed some comments, but it's not passing tests anymore, I have to see why. I'm just worrying about the fact of splitting the test functions might have created some implicit global state that I don't see. :/

**Comment by RaitoBezarius** [comment]
Also, it brings some question.

If I have some hunter depending on A & B

If I publish A, A, A, …, A, B, B, B, …, B

How many times should the hunter be called?

**Comment by iYehuda** [comment]
> Also, it brings some question.
> 
> If I have some hunter depending on A & B
> 
> If I publish A, A, A, …, A, B, B, B, …, B
> 
> How many times should the hunter be called?

That's a good point.
Let's examine the use case with an example:
* Wer'e running on Azure (so AzureMetadataAPI event is published)
* There are 2 vulnerable kubelet APIs (so SecureKubeletEvent is published twice)
Assume there's a new vulnerability in azure that depends on the former combination, we would like to get it triggered for every kubelet found but we need only one metadata api as prerequisite.

It is possible to have a case which additional context should be included in triggering, so supplying a predicate for two events can be done to filter context.
In such case, you might have event A published twice and event B published twice as well, but a registered hunter C might need to have the second A and first B.
This way, you'll need to trigger C for each combination of A and B events and let it filter.

**Comment by danielsagi** [comment]
Closing due to inactivity

**Pull request closed without merging**